### PR TITLE
Write guides to implement new-style functions

### PR DESCRIPTION
**Pull request by kmaehashi**
I started to write docs on how to implement user-defined new-style functions.

I'd like to discuss two points:

## 1. How do we reorganize this docs?

Currently I'm planning to describe this way:

> In the following sections, we first introduce steps to implenent an old-style function. After that we describe the difference of implementation between old-style functions and new-style functions.

Or, should we completely rewrite the guide to use new-style functions?

I think old-style functions are still convenient for many people, as the implementation cost is less than new-style, so I'd like to proceed with the current style.

## 2. Can we give some better name to "old-style" functions?

I think the name "old-style" make users feel "it is not recommended to use".
How about giving less-aggressive name, like "classic-style" or "conventional"?

### CODE REVIEW

**Comment by kmaehashi** [comment]
Refs #4449.
This document also intends to cover topics from https://gist.github.com/beam2d/dd87d97b8d7581f95cd19b1de86c9f4d.

**Comment by niboshi** [comment]
>I think old-style functions are still convenient for many people, as the implementation cost is less than new-style, so I'd like to proceed with the current style.

Do you mean when optimized implementation is required?
Do normal users usually need optimized implementation?

>Can we give some better name to "old-style" functions?

How about "functions with/without higher-order backpropagation"?

**Comment by kmaehashi** [comment]
As for users, it's simple that functions can simply implemented in one class using NumPy/CuPy. (like done in other frameworks).
I'm not sure if higher-order backprop will widely be used in many other domains. In that case I think it's the right direction (in terms of designing the ecosystem) to always recommend users to use FunctionNode so that community assets will become more valuable for longer period.

As for the second point, I overlooked that the terms are already defined here (Notes box), so I'll proceed with the current terms.
https://docs.chainer.org/en/stable/reference/generated/chainer.FunctionNode.html#chainer.FunctionNode

**Comment by kmaehashi** [comment]
We discussed offline:

* We recommend users to use FunctionNode in the first place.
* However, still leave documentation for Function (expected readers are users who needs to read old code)
* Also add guidance to help users implement functions without having to write backward (i.e., "non-double-backpropable" Function) to reduce implementation cost.
  (It may be better to have a built-in feature for this; this point is out of scope of this PR.)

**Comment by toslunar** [comment]
I expect this PR closes #5707.

File: docs/source/guides/functions.rst [hunk 1]
```diff
@@ -17,13 +17,38 @@ After reading this section, you will be able to:
...
+
+In Chainer, you can define a function in two ways: old-style and new-style.
+
+* Old-style functions inherit from :class:`chainer.Function` class.
+  Forward and backward computation can be implemented using NumPy/CuPy.
+* New-style functions inherit from :class:`chainer.FunctionNode` class (introduced in Chainer v3).
+  Forward computation can be implemented using NumPy/CuPy.
+  Backward computation needs to be implemented using other new-style functions.
+
+The primary advantage of using new-style functions is that they support computation of higher-order gradient (a.k.a. higher-order derivative or double back-propagation).
```
Comment (Crissman) [inline docs/source/guides/functions.rst]: gradients

File: docs/source/guides/functions.rst [hunk 2]
```diff
@@ -17,13 +17,38 @@ After reading this section, you will be able to:
...
+Higher-order gradient is used in some models e.g., recently-proposed GAN architectures.
```
Comment (Crissman) [inline docs/source/guides/functions.rst]: gradients are used

File: docs/source/guides/functions.rst [hunk 3]
```diff
@@ -17,13 +17,38 @@ After reading this section, you will be able to:
...
+
+Currently, most of built-in functions are implemented as :class:`~chainer.FunctionNode` (with a few exceptions listed in `#4449 <https://github.com/chainer/chainer/issues/4449>`__).
+
+Basically, we suggest you to use new-style when implementing new functions.
```
Comment (Crissman) [inline docs/source/guides/functions.rst]: We recommend you use

File: docs/source/guides/functions.rst (hunk 1 above)
Comment (Crissman) [inline docs/source/guides/functions.rst]:
```suggestion
The primary advantage of using new-style functions is that they support computation of higher-order gradients (a.k.a. higher-order derivative or double back-propagation).
```

File: docs/source/guides/functions.rst (hunk 2 above)
Comment (Crissman) [inline docs/source/guides/functions.rst]:
```suggestion
Higher-order gradients are used in some models e.g., recently-proposed GAN architectures.
```

File: docs/source/guides/functions.rst (hunk 3 above)
Comment (Crissman) [inline docs/source/guides/functions.rst]:
```suggestion
Basically, we recommend you use new-style when implementing new functions.
```

**Comment by kmaehashi** [comment]
imosCI, test this please.

**Comment by kmaehashi** [comment]
I've finished working on this PR. Ready for another review.

File: docs/source/guides/functions.rst [hunk 4]
```diff
@@ -420,143 +485,38 @@ You can refer to :ref:`configuration` to see how to configure this flag as well
...
-
-.. testcode::
+We can fix it as follows::
 
   def dropout(x):
       if not chainer.config.train:
           return x
 
-      xp = backend.get_array_module(x.array)
+      xp = backend.get_array_module(x.data)
```
Comment (toslunar) [inline docs/source/guides/functions.rst]:
```suggestion
      xp = backend.get_array_module(x.array)
```

File: docs/source/guides/functions.rst [hunk 5]
```diff
@@ -420,143 +485,38 @@ You can refer to :ref:`configuration` to see how to configure this flag as well
...
 You will need to check the value of the boolean flag ``chainer.config.train`` and branch appropriately.
 
-For example, consider the following simple dropout function:
-
-.. testcode::
+For example, consider the following simple dropout function::
 
   def dropout(x):
-      xp = backend.get_array_module(x.array)
+      xp = backend.get_array_module(x.data)
```
Comment (toslunar) [inline docs/source/guides/functions.rst]:
```suggestion
      xp = backend.get_array_module(x.array)
```

**Comment by kmaehashi** [comment]
Jenkins, test this please.
